### PR TITLE
feat: permission based feature access

### PR DESCRIPTION
**Pull request by HamzaIbnFarooq**
#### Pre-Flight checklist
- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots

#### What are the relevant tickets?
https://github.com/mitodl/edx-sysadmin/issues/16

#### What's this PR do?
It adds permission-based feature access in which global staff user (is_staff) has access to all the features of sysadmin while course-admins (instructors) are bound to see only the git-logs of their courses.

We were trying to allow Course-Admins to update their courses from git using the Git-Import feature while restricting them from creating a new course using the same feature but that's not possible because until the course repo is fetched and the course is created we can't tell if the user has entered a new git repo link or updated an existing course. Thus we have to restrict Course-Admins from the whole Git-Import feature.

#### How should this be manually tested?
Try to access sysadmin with a course-admin user and an is_staff user and observe if they have the appropriate rights.

#### Screenshots (if appropriate)
Staff User (is_staff) view:
image

Course Admin view:
image

### CODE REVIEW

File: edx_sysadmin/views.py [hunk 1]
```diff
@@ -34,44 +33,88 @@
...
+            return reverse("sysadmin:gitlogs")
+        elif user_has_access_to_git_import_panel(self.request.user):
+            return reverse("sysadmin:gitimport")
+        else:
+            raise Http404
+
+
 @method_decorator(ensure_csrf_cookie, name="dispatch")
-@method_decorator(staff_member_required, name="dispatch")
+@method_decorator(user_has_sysadmin_access, name="dispatch")
```
Comment (ziafazal) [inline edx_sysadmin/views.py:67]: We should use `user_passes_test` decorator here and pass that decorator our custom validation method `user_is_staff` or `user_has_course_access` and put the logic of required permissions in those methods.

File: edx_sysadmin/views.py [hunk 2]
```diff
@@ -34,44 +33,88 @@
...
+    user_has_access_to_git_import_panel,
 )
+from edx_sysadmin.utils.decorators import user_has_sysadmin_access
+
 
 log = logging.getLogger(__name__)
 
 
+@method_decorator(user_has_sysadmin_access, name="dispatch")
+class SysadminDashboardRedirectionView(RedirectView):
```
Comment (ziafazal) [inline edx_sysadmin/views.py:48]: I don't think we need this view. `user_passes_test` decorator should be use on the sysadmin views and pass `login_url` and `redirect_field_name`.

**Review by ziafazal — COMMENTED** [review]
@HamzaIbnFarooq overall looks good. I have couple of suggestions.

File: edx_sysadmin/views.py (hunk 2 above)
Comment (HamzaIbnFarooq) [reply edx_sysadmin/views.py:48]: Sorry, it a bit confusing, let me explain my thoughts below to get this comment further:

Whenever a user hits `/sysadmin` this view checks User permissions and redirects the user to the first allowed views to which the user has access like `/sysadmin/users`, `/sysadmin/courses`, `/sysadmin/gitimport` and  `/sysadmin/gitlogs`.

Currently, if a user has access to both `users` and `courses` panel then by default the user will be redirected to `users` panel on hitting the sysadmin base URL `/sysadmin`. And when `user` hits `/courses` URL `intentionally`, he will be redirected to `courses` panel.

Keeping the mentioned flow in mind, we can't put the redirection logic in a decorator and use it on SysadminBaseView because if a user hits `/courses`, `the decorator will pick the first accessible panel and redirect the user to that`.

File: edx_sysadmin/views.py (hunk 1 above)
Comment (HamzaIbnFarooq) [reply edx_sysadmin/views.py]: I have updated the permission decorators.

File: .gitignore [hunk 3]
```diff
@@ -61,4 +61,5 @@ requirements/private.in
 requirements/private.txt
 
 # IDE related
-.idea/*
+.idea/
+.vscode/
```
Comment (pdpinch) [inline .gitignore:65]: Can you move this to a separate PR or commit?

Comment (HamzaIbnFarooq) [reply .gitignore]: Removed from this PR.

File: edx_sysadmin/views.py (hunk 2 above)
Comment (ziafazal) [reply edx_sysadmin/views.py:48]: 👍

**Review by ziafazal — APPROVED** [review]
👍
